pkg/tfgen/parse/extension.go

```diff
@@ -85,3 +121,98 @@ func WalkNode[T ast.Node](node ast.Node, f func(T)) {
 	})
 	contract.AssertNoErrorf(err, "impossible: ast.Walk never returns an error")
 }
+
+type renderType struct {
+	kind ast.NodeKind
+	f    renderer.NodeRendererFunc
+}
+
+func (renderType renderType) RegisterFuncs(r renderer.NodeRendererFuncRegisterer) {
+	r.Register(renderType.kind, renderType.f)
+}
+
+func panicOnRender(writer util.BufWriter, source []byte, n ast.Node, entering bool) (ast.WalkStatus, error) {
+	contract.Failf("The renderer for %s should not have been called", n.Kind())
+	return ast.WalkStop, nil
+}
+
+var _ renderer.NodeRenderer = (*tableRenderer)(nil)
+
+type tableRenderer struct {
+	renderer.Renderer
+}
+
+func (tableRenderer tableRenderer) RegisterFuncs(r renderer.NodeRendererFuncRegisterer) {
+	r.Register(extensionast.KindTable, tableRenderer.render)
+	r.Register(extensionast.KindTableHeader, panicOnRender)
+	r.Register(extensionast.KindTableRow, panicOnRender)
+	r.Register(extensionast.KindTableCell,
+		func(writer util.BufWriter, source []byte, n ast.Node, entering bool) (ast.WalkStatus, error) {
+			return ast.WalkContinue, nil
+		})
+}
+
+func (tableRenderer tableRenderer) render(
+	writer util.BufWriter, source []byte, n ast.Node, entering bool) (ast.WalkStatus, error) {
+	if !entering {
+		return ast.WalkSkipChildren, nil
+	}
+	_, err := writer.WriteRune('\n') // this is so that we have a newline between markdown elements.
+	contract.AssertNoErrorf(err, "impossible")
+	var inHeader bool
+	header := make([]string, 0, len(n.(*extensionast.Table).Alignments))
+	var rows [][]string
+	err = ast.Walk(n, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
+		switch n := n.(type) {
+		case *extensionast.Table:
+			return ast.WalkContinue, nil
+		case *extensionast.TableHeader:
+			inHeader = entering
+			return ast.WalkContinue, nil
+		case *extensionast.TableRow:
+			if entering {
+				rows = append(rows, make([]string, 0, len(n.Alignments)))
+			}
+			return ast.WalkContinue, nil
+		case *extensionast.TableCell:
+			if entering {
+				var cell bytes.Buffer
+				textBlock := ast.NewTextBlock()
+				child := n.FirstChild()
+				for {
+					if child == nil {
+						break
+					}
+					next := child.NextSibling()
+					textBlock.AppendChild(textBlock, child)
+					child = next
+				}
+				err := tableRenderer.Render(&cell, source, textBlock)
+				if err != nil {
+					return ast.WalkStop, err
+				}
+				content := strings.TrimSpace(cell.String())
+				if inHeader {
+					header = append(header, content)
+				} else {
+					rows[len(rows)-1] = append(rows[len(rows)-1], content)
+				}
+			}
+			return ast.WalkSkipChildren, nil
+		default:
+			return ast.WalkStop, fmt.Errorf("unexpected node in a table: %s", n.Kind().String())
```

Comment:
I'm pretty sure that `%s` implies `.String()` when available.
```suggestion
			return ast.WalkStop, fmt.Errorf("unexpected node in a table: %s", n.Kind())
```